# Contributions of 5mC to Corvus evolution

This github release corresponding to Apr 2025 is archived on zenodo: [![DOI](https://zenodo.org/badge/656417446.svg)](https://doi.org/10.5281/zenodo.11244554)

Large data files can be found on zenodo: https://doi.org/10.5281/zenodo.15243953

**Highly recommended to open the script file `.md` in a markdown editor (typora/obsidian), or use the `.hmtl`.**
 
## Overview
This repository contains directories structured to facilitate a comprehensive analysis workflow integrating resequencing data and 3 5mC experiments.

### `genome_preparation`
Scripts and resources for preparing genome associated data for analysis (e.g. cut sites, annotation feature).

### `resequencing_reanalysis`
Contains tools and outputs for the reanalysis of resequenced data.

### `spatial_plotting`
Scripts for generating spatial distributions.

### `5mC_processing`
Processing of 5mC data, including pre-processing, QC, and generation of bismark `cov.gz` files.

### `decision_trees`
Includes machine learning implementations of random forest and boosted regression trees to assess chromosomal determinants of 5mC. 

### `methylomic_divergence`
Scripts for analysis on methylomic divergence - largely manhattan plots. 

### `ordinations`
Scripts for performing ordination methods to analyze multivariate 5mC and resequencing data. 

---

## Contact

For any questions or concerns, please open an issue on this repository or reach out to Justin Merondun: heritabilities@gmail.com

